### PR TITLE
Migrate bokehjs' visual baselines to LFS

### PR DESCRIPTION
**Pull request by mattpap**
I found myself again the previous and this week postponing PRs and working around large scale changes to visual baselines, to avoid inflating the repository size unnecessarily. After going through the available alternatives to storing baselines in the repository, I came to conclusion that LFS is the only reasonable alternative.

Though LFS requires an external tool, i.e. `git-lfs`, it's an easy install and only required when dealing with visual baselines. Cloning doesn't require `git-lfs` to be installed and baselines can be pulled at a later time (`git lfs pull`) if needed. We already pay for LFS storage and bandwidth, so no additional setup is required. Given that currently visual baselines cumulatively only weigh about 20 MB, we shouldn't need to increase our quota. 

The only disadvantage I found so far is increased testing time, about 15%~20% locally but quite significantly higher in CI. However, the existing approach of calling git a thousand or so times is already slow and this just makes it slower. There should be a better approach to do this, which perhaps could make this faster and it currently is. So this can be a good motivation to revisit how we resolve baselines.

TODO:

- [x] update docker workflow
- [x] update documentation

fixes #11190

### CODE REVIEW

File: scripts/ci/install_node_modules.sh [hunk 1]
```diff
@@ -5,5 +5,4 @@ set -x #echo on
 cd bokehjs
 npm install --location=global npm
 npm ci --no-progress
-node make examples --no-build
-git restore package*.json # avoid .dirty in the package version
+node make --no-up-to-date examples:build
```
Comment (mattpap) [inline scripts/ci/install_node_modules.sh]: This now fails. Do we expect bokehjs to be already built at this point or not? `--no-build` didn't do anything for awhile now, so bokehjs was built here regardless.

**Comment by ianthomas23** [comment]
I've tried this out locally and it works seamlessly but is painfully slow.

On `branch-3.4` if I run `time node make test:integration` it takes 2.5 minutes. With this PR branch it takes over 24 minutes. This is on a ~2 year old mid-spec HP laptop running Ubuntu 22.04.

**Comment by mattpap** [comment]
> With this PR branch it takes over 24 minutes.

Yikes, that's worse than in CI. Then definitely this PR needs performance issues resolved.

File: scripts/ci/install_node_modules.sh (hunk 1 above)
Comment (mattpap) [reply scripts/ci/install_node_modules.sh]: It looks like this was unnecessary altogether.

**Comment by mattpap** [comment]
@ianthomas23, can you try it again? At least locally and in CI performance is back to normal or even improved by ~10%.

**Comment by ianthomas23** [comment]
> @ianthomas23, can you try it again? At least locally and in CI performance is back to normal or even improved by ~10%.

I'll try it out first thing tomorrow and report back.

**Comment by bryevdv** [comment]
In this PR images now show
> Unable to render rich display
> 
> Invalid image source.

Just to set future expectations: when baselines change will it still be possible to easily view a comparison of the old/new images anywhere?

**Comment by mattpap** [comment]
> Just to set future expectations: when baselines change will it still be possible to easily view a comparison of the old/new images anywhere?

Based on [this](https://docs.github.com/en/repositories/working-with-files/managing-large-files/collaboration-with-git-large-file-storage#viewing-large-files-in-pull-requests), I'm inferring that sadly it isn't possible (at least for now). Being able to preview new and changed baselines has been really helpful.

**Comment by ianthomas23** [comment]
> @ianthomas23, can you try it again? At least locally and in CI performance is back to normal or even improved by ~10%.

Performance is definitely better for me locally, 15% faster than `branch-3.4`. Very nice!

> > Just to set future expectations: when baselines change will it still be possible to easily view a comparison of the old/new images anywhere?
> 
> Based on [this](https://docs.github.com/en/repositories/working-with-files/managing-large-files/collaboration-with-git-large-file-storage#viewing-large-files-in-pull-requests), I'm inferring that sadly it isn't possible (at least for now). Being able to preview new and changed baselines has been really helpful.

I suppose this is a blocker as it will make reviewing PRs much harder than it is currently. I did find out how to fix it locally so that VSCode before-and-after image comparison works by using `git config diff.lfs.textconv cat` (ref microsoft/vscode#86611 and links therein).

**Comment by mattpap** [comment]
> I suppose this is a blocker as it will make reviewing PRs much harder than it is currently.

I view this as more as an annoyance than a blocker, and a little price to pay to resolve the current situation. With LFS there's at least hope that GitHub could implement preview, because I don't think it's any harder than just to detect a file is an LFS pointer and resolve it. I will be looking into third-party services that could restore this functionality, or maybe roll out something of my own. I really would like to have some way of previewing test reports.

**Comment by mattpap** [comment]
@bokeh/core, I would like to get a resolution on this. To me the only real alternative to this PR, is to keep baselines in the repository and stop caring about repository size and perhaps at some point filter the repository to remove "old" images.

**Comment by bryevdv** [comment]
Is there a way to track and plot the repository growth over time? Not having image diffs in PR is a serious drawback. I am now looking for a reason to keep the status quo. I.e if the growth is large then we will have to learn to live with this drawback, I guess, but if the growth is relatively slow the I could probably be placated into non-action here. 

Alternatively, are there any issues or discussions on GH support regarding this? If not, we could make one ourselves in any case. But it would be useful to know if there has already been any discussion or response. 

Do we know if image diffs would show up if the images were stored in a sub-repo instead? I don't love sub-repos but our process is streamlined enough now that I could imagine introducing a little complexity, if that's what it takes to obtain a good outcome.

**Comment by mattpap** [comment]
If my calculations are correct then the first plot shows the size of `bokehjs/test/baselines/linux` for each commit since introduction of this directory (`*.png` files only, because `*.blf` are insignificant in size). The second plot shows a histogram of how much each commit contributes and a cumulative size of baseline contributions (the total is 66 MB).
image
Note this doesn't include history before `linux` subdirectory, but earlier contributions are insignificant. This also doesn't include osx and windows baselines, which effectively tripled the size of each contribution for that time interval.

**Comment by bryevdv** [comment]
GH recommendation on size is:

> We recommend repositories remain small, ideally less than 1 GB, and less than 5 GB is strongly recommended. 

Given those guidelines and the chart above, I think we have some breathing room. I'm also not opposed *in principle* to filtering the repo at some point, despite some disruption that would cause, either. I would vote to wait on this for now, and keep periodically trying to see if some path to LFS + PR image diffs can be found in the next few years.

**Comment by mattpap** [comment]
I agree with the previous comment, i.e. to keep baselines in the repository and update them freely, and revisit this periodically. @bokeh/core, unless there are other comments or concerns, I'm going to decline and close this PR. I will follow up with another PR, which will extract useful changes and especially any performance improvements to the testing infrastructure, which is unrelated to LFS.

**Pull request closed without merging**